artemis/modules/wordpress_plugins.py

```diff
@@ -0,0 +1,300 @@
+
+            if previous_line == "changelog" or (has_reversed_changelog and seen_changelog_line):
+                seen_changelog_line = True
+                # Some changelog entries have the format <slug>: <version>
+                if line.startswith(slug):
+                    line = line[len(slug) :].strip(" :")
+                # Some changelog entries have the format version <version>
+                if "version" in line:
+                    line = line[line.find("version") + len("version") :].strip(" :")
+                # Some changelog entries have the format V <version>
+                if "v " in line:
+                    line = line[line.find("v ") + len("v ") :].strip(" :")
+                version = (
+                    line.replace("(", " ")
+                    .replace("*", " ")
+                    .replace("[", " ")
+                    .replace("]", " ")
+                    .replace("'", " ")
+                    .replace(":", " ")
+                    .replace(",", " ")
+                    .replace("-", " ")
+                    .replace("<h4>", " ")
+                    .replace("</h4>", " ")
+                    .strip()
+                    # Some versions are prefixed with 'v' (e.g. v1.0.0)
+                    .lstrip("v")
+                    .split(" ")[0]
+                )
+                if "." in version:
+                    changelog_version = version
+                    if not has_reversed_changelog:
+                        break
+
+            previous_line = line
+
+    # Some plugins have broken "stable tag"
+    if slug in PLUGINS_TO_SKIP_STABLE_TAG:
+        return changelog_version
+
+    tag_lines = [line for line in readme_content.lower().split("\n") if line.strip("* ").startswith("stable tag:")]
+    if len(tag_lines) > 1:
+        return changelog_version
+
+    if len(tag_lines):
+        (tag_line,) = tag_lines
+        tag = tag_line.strip().split(":")[1].strip()
+
+        if tag != "trunk" and (
+            # Sometimes the changelog version is greater, sometimes the "stable tag" version is greater -
+            # let's pick the greater one as the version.
+            changelog_version is None
+            or tuple(tag.split(".")) > tuple(changelog_version.split("."))
+        ):
+            return tag
+    return changelog_version
+
+
+def _get_host_from_url(url: str) -> str:
+    host = urllib.parse.urlparse(url).hostname
+    assert host is not None
+    return host
+
+
+def _get_plugins_from_homepage(url: str) -> List[Dict[str, Any]]:
+    url_parsed = urllib.parse.urlparse(url)
+    response = http_requests.get(url)
+    soup = BeautifulSoup(response.text)
+    links = []
+    for tag in soup.find_all():
+        new_url = None
+        for attribute in ["src", "href"]:
+            if attribute not in tag.attrs:
+                continue
+
+            new_url = urllib.parse.urljoin(url, tag[attribute])
+            new_url_parsed = urllib.parse.urlparse(new_url)
```

Comment:
this part was used in finding related links to scanned page. I would make it metafunction, so there won't be repetitions